### PR TITLE
etcdserver: add trace for txn request

### PR DESCRIPTION
**Pull request by YoyinZyc**
Here is an example output of a txn request. 
Request:
```
cmp := clientv3.Compare(clientv3.Value("foo"), "=", "bar");
op1 := clientv3.OpGet("foo", clientv3.WithRange("fooo"))
op2 := clientv3.OpPut("txn", "put")
op3 := clientv3.OpPut("put","txn")
op4 := clientv3.OpTxn([]clientv3.Cmp{cmp}, []clientv3.Op{op1}, []clientv3.Op{})
txn := cli.Txn(ctx).If(cmp).Then(op1, op2, op3, op4)
```
Output:
```
{"level":"info","ts":"2020-01-02T15:27:52.215-0800","caller":"traceutil/trace.go:166","msg":"trace[1023984242] transaction",
"detail":"{read_only:false; response_revision:7; number_of_response:4; }","duration":"440.713µs","start":"2020-01-02T15:27:52.215-0800","end":"2020-01-02T15:27:52.215-0800",
"steps":["trace[1023984242] 'process raft request'  (duration: 195.497µs)",
"trace[1023984242] 'compare'  (duration: 36.258µs)","trace[1023984242] 'check requests'  (duration: 1.317µs)",
"trace[1023984242] 'range keys from in-memory index tree' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 12.56µs)",
"trace[1023984242] 'range keys from bolt db' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 2.126µs)",
"trace[1023984242] 'filter and sort the key-value pairs' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 1.034µs)",
"trace[1023984242] 'assemble the response' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 308ns)",
"trace[1023984242] 'get key's previous created_revision and leaseID' {req_type:put; key:txn; req_size:10; } (duration: 9.659µs)",
"trace[1023984242] 'marshal mvccpb.KeyValue' {req_type:put; key:txn; req_size:10; } (duration: 1.766µs)",
"trace[1023984242] 'store kv pair into bolt db' {req_type:put; key:txn; req_size:10; } (duration: 9.589µs)",
"trace[1023984242] 'attach lease to kv pair' {req_type:put; key:txn; req_size:10; } (duration: 61ns)",
"trace[1023984242] 'get key's previous created_revision and leaseID' {req_type:put; key:put; req_size:10; } (duration: 1.192µs)",
"trace[1023984242] 'marshal mvccpb.KeyValue' {req_type:put; key:put; req_size:10; } (duration: 306ns)",
"trace[1023984242] 'store kv pair into bolt db' {req_type:put; key:put; req_size:10; } (duration: 1.139µs)",
"trace[1023984242] 'attach lease to kv pair' {req_type:put; key:put; req_size:10; } (duration: 47ns)",
"trace[1023984242] 'range keys from in-memory index tree' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 2.293µs)",
"trace[1023984242] 'range keys from bolt db' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 897ns)",
"trace[1023984242] 'filter and sort the key-value pairs' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 146ns)",
"trace[1023984242] 'assemble the response' {req_type:range; range_begin:foo; range_end:fooo; } (duration: 112ns)"],
"step_count":19}
```
This example sets the `traceThreshold ` to 0 to show every step of the request. Normally, the step will be logged out only when the duration exceeds 100ms.
https://github.com/etcd-io/etcd/blob/e6980b1f9fc008837abb9177e87e893e48d565c1/etcdserver/v3_server.go#L42

Here are the benchmark results from 10000 times run of the above example which will trigger trace logging.
Without trace:
image
With trace:
image

### CODE REVIEW

**Comment by YoyinZyc** [comment]
@gyuho @xiang90 Please take a look. Thanks.

File: etcdserver/apply.go [hunk 1]
```diff
@@ -216,13 +220,13 @@ func (a *applierV3backend) Put(txn mvcc.TxnWrite, p *pb.PutRequest) (resp *pb.Pu
 
 	var rr *mvcc.RangeResult
 	if p.IgnoreValue || p.IgnoreLease || p.PrevKv {
-		trace.DisableStep()
-		rr, err = txn.Range(p.Key, nil, mvcc.RangeOptions{})
+		trace.StepWithFunction(func() {
+			rr, err = txn.Range(p.Key, nil, mvcc.RangeOptions{})
+		}, "get previous kv pair")
+
```
Comment (gyuho) [inline etcdserver/apply.go]: why are these changes needed?

Comment (gyuho) [reply etcdserver/apply.go]: nvm, I see `StepWithFunction` is being added below.

File: pkg/traceutil/trace.go [hunk 2]
```diff
@@ -167,6 +195,17 @@ func (t *Trace) logInfo(threshold time.Duration) (string, []zap.Field) {
...
-		zap.Strings("steps", steps)}
+		zap.Strings("steps", steps),
+		zap.Int("step_count", len(t.steps))}
 	return msg, fs
 }
+
+func (t *Trace) hasField(f Field) bool {
+	for i, v := range t.fields {
+		if v.Key == f.Key {
+			t.fields[i].Value = f.Value
```
Comment (xiang90) [inline pkg/traceutil/trace.go]: if we update this trace field value here, the func naming needs to be changed. hasField should be a func name for a read only operation.

Comment (xiang90) [reply pkg/traceutil/trace.go]: probably use: updateFieldIfExist instead.

**Comment by xiang90** [comment]
how much would this affect performance?

**Comment by xiang90** [comment]
lgtm in general

File: pkg/traceutil/trace.go (hunk 2 above)
Comment (YoyinZyc) [reply pkg/traceutil/trace.go]: Fixed. Thanks for pointing out.

**Comment by YoyinZyc** [comment]
> how much would this affect performance?

Added benchmark results above.

File: pkg/traceutil/trace.go [hunk 3]
```diff
@@ -102,13 +104,26 @@ func (t *Trace) InsertStep(at int, time time.Time, msg string, fields ...Field)
...
+
 // Step adds step to trace
 func (t *Trace) Step(msg string, fields ...Field) {
 	if !t.stepDisabled {
 		t.steps = append(t.steps, step{time: time.Now(), msg: msg, fields: fields})
 	}
 }
 
+// StepWithFunction will measure the input function as a single step
+func (t *Trace) StepWithFunction(f func(), msg string, fields ...Field) {
```
Comment (jingyih) [inline pkg/traceutil/trace.go]: I like the idea of `StepWithFunction`, do we still need to expose `EnableStep` and `DisableStep`?

File: pkg/traceutil/trace.go [hunk 4]
```diff
@@ -102,13 +104,26 @@ func (t *Trace) InsertStep(at int, time time.Time, msg string, fields ...Field)
 	}
 }
 
+// SubTrace adds step to trace as a sublevel trace
+func (t *Trace) SubTrace(fields ...Field) {
```
Comment (jingyih) [inline pkg/traceutil/trace.go]: Does this also count towards the final "step count"?

Comment (jingyih) [reply pkg/traceutil/trace.go]: Also, can we provide a unit test or example to indicate the usage and expected behavior of this function?

File: pkg/traceutil/trace.go (hunk 3 above)
Comment (YoyinZyc) [reply pkg/traceutil/trace.go]: Yes you are right. I changed them to private functions.

File: pkg/traceutil/trace.go (hunk 4 above)
Comment (YoyinZyc) [reply pkg/traceutil/trace.go]: > Does this also count towards the final "step count"?

It should not count into `step_count`. I fixed it to `len(step)`


> Also, can we provide a unit test or example to indicate the usage and expected behavior of this function?

Added one test case in `TestLog`

**Comment by YoyinZyc** [comment]
Friendly ping. I rebased this PR. Is there anything I need to change or can we have it approved? Thanks.
@gyuho @xiang90 @jingyih

**Comment by xiang90** [comment]
LGTM
